craft_application/util/yaml.py

```diff
@@ -92,3 +102,38 @@ def safe_yaml_load(stream: TextIO) -> Any:  # noqa: ANN401 - The YAML could be a
     # Silencing S506 ("probable use of unsafe loader") because we override it by using
     # our own safe loader.
     return yaml.load(stream, Loader=_SafeYamlLoader)  # noqa: S506
+
+
+@overload
+def dump_yaml(
+    data: Any, stream: TextIO, **kwargs: Any  # noqa: ANN401 Any gets passed to pyyaml
+) -> None:
+    ...  # pragma: no cover
+
+
+@overload
+def dump_yaml(
+    data: Any,  # noqa: ANN401 Any gets passed to pyyaml
+    stream: None = None,
+    **kwargs: Any,  # noqa: ANN401 Any gets passed to pyyaml
+) -> str:
+    ...  # pragma: no cover
+
+
+def dump_yaml(data: Any, stream: TextIO | None = None, **kwargs: Any) -> str | None:
+    """Dump an object to YAML using PyYAML.
+
+    This works as a drop-in replacement for ``yaml.safe_dump``, but adjusting
+    formatting as appropriate.
+
+    :param data: the data structure to dump.
+    :param stream: The optional text stream to which to write.
+    :param kwargs: Keyword arguments passed to pyyaml
+    """
+    yaml.add_representer(
+        str, _repr_str, Dumper=cast(Type[yaml.Dumper], yaml.SafeDumper)
+    )
+    kwargs.setdefault("sort_keys", False)
+    return cast(  # This cast is needed for pyright but not mypy
+        Union[str, None], yaml.dump(data, stream, Dumper=yaml.SafeDumper, **kwargs)
+    )
```

Comment:
Why 3 `dump_yaml` here?

Reply:
Overloads for type checkers. The first one says that it won't return anything if you give it a stream (because it writes to that stream as a side-effect instead). The second says if you don't give it a stream it returns a string. The non-overload one is the actual implementation.

I've been contributing to typeshed recently so I'm getting more into stricter typing annotations.